### PR TITLE
feat(simple_packing): auto-compute encoder params + sp_* descriptor key rename

### PR DESCRIPTION
**Pull request by tlmquintino**
## Summary

Two paired changes that together drop the explicit
`compute_packing_params` step from `simple_packing` user code and
align the codec's descriptor keys with the rest of the codebase
convention.

```python
# Before
params = tensogram.compute_packing_params(temps.ravel(), bits_per_value=16, decimal_scale_factor=0)
desc = {"type": "ntensor", "shape": [...], "dtype": "float64",
        "encoding": "simple_packing", "compression": "zstd", **params}

# After
desc = {"type": "ntensor", "shape": [...], "dtype": "float64",
        "encoding": "simple_packing", "sp_bits_per_value": 16,
        "compression": "zstd"}
tensogram.encode(meta, [(desc, data)])  # the encoder auto-computes the rest
```

## Added

- **Auto-compute path** in the encoder: when the descriptor carries
  `encoding: "simple_packing"` + `sp_bits_per_value` and the derived
  keys (`sp_reference_value` / `sp_binary_scale_factor`) are absent,
  the encoder calls `simple_packing::compute_params` on the input
  data and stamps the full four-key set into the descriptor.  The
  encoded message stays self-describing — decoders see no difference
  from a hand-written explicit descriptor.
- **Wire-format ergonomics**: `sp_bits_per_value` is required;
  `sp_decimal_scale_factor` defaults to `0` when absent; explicit
  `sp_reference_value` + `sp_binary_scale_factor` win verbatim
  (advanced workflows that pin the reference value across
  time-series).
- **`StreamingEncoder::write_object`** uses the same auto-compute
  hook.  `write_object_pre_encoded` is unchanged — pre-encoded bytes
  are opaque, the user must supply the full four-key set.
- 14 Rust BDD scenarios in `rust/tensogram/tests/simple_packing_auto_compute.rs`
  (S1–S7 design-plan + parity check + E1–E5 hardening).
- 11 mirrored Python BDD scenarios in `python/tests/test_simple_packing_auto_compute.py`.
- 2 new C++ scenarios in `cpp/tests/test_simple_packing.cpp`
  (`AutoComputeFromBitsOnly` and `HalfExplicitParamsRejected`).
- Examples (`examples/python/03_simple_packing.py`, Rust + C++ twins,
  Jupyter notebooks) lead with the ergonomic form and keep the
  explicit form as a secondary advanced demo.

## Changed

- **Wire format**: the four `simple_packing` descriptor keys have been
  renamed to match the `<codec>_<key>` convention used by every other
  codec (`szip_*`, `zstd_*`, `blosc2_*`, `zfp_*`, `shuffle_*`):

  | Old                  | New                      |
  |----------------------|--------------------------|
  | `bits_per_value`     | `sp_bits_per_value`      |
  | `decimal_scale_factor`| `sp_decimal_scale_factor`|
  | `reference_value`    | `sp_reference_value`     |
  | `binary_scale_factor`| `sp_binary_scale_factor` |

  Preamble wire version stays at 3.  Pre-rename v3 messages with
  unprefixed `simple_packing` keys are not readable (pre-public
  software, no deprecation path).

- `compute_packing_params` (Python / WASM / C FFI / TypeScript) now
  returns a dict / object with `sp_*`-prefixed keys, so callers can
  spread directly into a descriptor: `desc = {..., **params}`.

- `convert-grib` and `convert-netcdf` automatically emit the new keys
  via the centralised `pipeline::apply_pipeline` helper — no per-converter
  edits needed.

- `python/tensogram-anemoi` switched to the auto-compute form
  (drops the per-field `compute_packing_params` call).

- Documentation: `docs/src/encodings/simple-packing.md` gains a
  prominent "Quick start — auto-compute" section showing the
  ergonomic form in Python, Rust, and TypeScript; explicit form
  retained for advanced use.

## Fixed

- **Half-explicit params silently overwritten** (Pass-3 hardening):
  providing exactly one of `sp_reference_value` /
  `sp_binary_scale_factor` (and not the other) used to silently get
  overwritten by auto-compute — a precision footgun.  Now rejected
  with a clear error naming both keys.

- **Auto-compute on substituted bytes distorted `sp_reference_value`**
  (Pass-3 hardening): when `allow_nan=true` substituted NaN with 0.0,
  auto-compute over the substituted bytes inflated the min/max range
  (e.g. [270, 285] → [0, 285]), silently halving precision at
  16 bits.  Auto-compute now runs on the **original** `data`, so
  non-finite inputs surface as a clean `PackingError`.

## Verification

Pre-flight checks all green:

| Layer | Result |
|-------|--------|
| `cargo fmt --check` | clean |
| `cargo clippy --workspace --all-targets -- -D warnings` | clean |
| `cargo clippy -p tensogram --features "remote,async" -- -D warnings` | clean |
| `cargo test --workspace` | all green |
| `cargo test -p tensogram --features "remote,async"` | all green |
| `cargo build --workspace` | all green |
| `ruff check` + `ruff format --check` on `python/tests/` | clean |
| `pytest python/tests/` | 541 passed (529 + 12 remote) |
| `pytest python/tensogram-xarray/tests/` | 242 passed |
| `pytest python/tensogram-zarr/tests/` | 235 passed |
| `pytest python/tensogram-anemoi/tests/` | 18 passed |
| `ctest --test-dir build` (C++) | 146 passed (incl. 8 simple_packing) |
| `make ts-test` | 376 passed (27 files) |
| `wasm-pack test --node` | 161 passed |
| `mdbook build docs/` | clean |
| `examples/python/03_simple_packing.py` | runs end-to-end |
| `cargo run --bin 03_simple_packing` | runs end-to-end |

## Notes

- **No version bump** in this PR (per user instruction; will be done
  at release time).  Wire version stays at 3.
- **Quality gates** — branch went through Pass 1-4 of the project's
  `/make-further-pass` strictness ladder before this PR was opened
  (commits `53a77e26`, `c5906a14`, `7f24c2f1`, `d89b1a2d`).
- **Scope cleanly separated** — the merged-in `chore/opencode-command-frontmatter`
  branch (commit `c8ec3fca`) was a small infrastructure fix for opencode
  command discovery, already merged into `main` via #96; merging it
  in here was a transient convenience while both branches were live
  and now collapses cleanly against main.

## Branch history

```
d89b1a2d  test(simple_packing): drop redundant `mut` flagged by clippy --all-targets
7f24c2f1  test(simple_packing): pass-4 hardening — pipeline / byte-order / multi-object / C++ parity
c5906a14  refactor(simple_packing): pass-3 hardening — edge cases + version-agnostic docs
53a77e26  refactor(simple_packing): pass-2 cleanup — reuse existing param helpers
1ad6d165  Merge branch 'chore/opencode-command-frontmatter' into feature/simple-packing-auto-compute
c8ec3fca  chore(opencode): add YAML frontmatter to project commands  ← already in main via #96
e87cc696  refactor(anemoi): use simple_packing auto-compute (Cycle 9)
75d1797d  docs(simple_packing): docs + examples + CHANGELOG + plans/DONE (Cycle 8)
b20d0b42  refactor(simple_packing): sp_* key rename in C++ / TS / examples (Cycle 6)
5b057c21  feat(simple_packing, py): auto-compute tests + key renames (Cycle 4)
ea62ea46  feat(simple_packing): encoder auto-computes sp_reference_value + sp_binary_scale_factor (Cycle 2)
d0e25546  refactor(simple_packing): rename descriptor keys to sp_* prefix (Cycle 1)
```

Docs Preview
https://sites.ecmwf.int/docs/tensogram/pull-requests/PR-97

### CODE REVIEW

File: rust/tensogram/src/encode.rs [hunk 1]
```diff
@@ -907,26 +924,170 @@ pub(crate) fn build_pipeline_config_with_backend(
...
+        .map(|c| {
+            let mut buf = [0u8; 8];
+            buf.copy_from_slice(c);
+            match byte_order {
+                ByteOrder::Big => f64::from_be_bytes(buf),
+                ByteOrder::Little => f64::from_le_bytes(buf),
+            }
+        })
+        .collect())
+}
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram/src/encode.rs:1116]: Fixed in 69451300. `bytes_as_f64_vec` now uses `try_reserve_exact` + a push loop, mirroring the `tensogram_encodings::pipeline::bytes_to_f64` pattern. Allocation failure surfaces as `TensogramError::Encoding` carrying the requested byte count, so encode-time auto-compute under memory pressure stays fallible instead of aborting.

File: rust/tensogram/src/encode.rs [hunk 2]
```diff
@@ -907,26 +924,170 @@ pub(crate) fn build_pipeline_config_with_backend(
...
+    // Explicit computed params present — trust them, skip the
+    // auto-compute work entirely.  We still default the
+    // sp_decimal_scale_factor knob when absent so the pipeline's
+    // extract_simple_packing_params doesn't fault on a missing key.
+    if has_ref && has_bsf {
+        desc.params
+            .entry("sp_decimal_scale_factor".to_string())
+            .or_insert(ciborium::Value::Integer(0i64.into()));
+        return Ok(());
+    }
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram/src/encode.rs:1049]: Fixed in 69451300. The `sp_bits_per_value` presence check is now hoisted ahead of the explicit-params short-circuit so the auto-compute-aware diagnostic fires uniformly on every code path — pure auto-compute, ref+bsf-only, and the full explicit 4-key form. Added regression test `s3b_explicit_path_missing_bits_per_value_is_a_clear_error` (parallel to the pre-existing `s3` for the auto-compute path) and updated the older `edge_cases::missing_required_param` to assert on the new key-named message rather than the generic "missing required" string from the low-level accessor.

File: rust/tensogram/src/encode.rs [hunk 3]
```diff
@@ -907,26 +924,170 @@ pub(crate) fn build_pipeline_config_with_backend(
...
+
+    // simple_packing only supports float64.  Checked here with an
+    // auto-compute-aware message; `build_pipeline_config` checks again
+    // for the explicit-params path.
+    if desc.dtype.byte_width() != 8 {
+        return Err(TensogramError::Encoding(format!(
+            "simple_packing only supports float64 dtype; got {:?}",
+            desc.dtype
+        )));
+    }
```
[the remark this answers is not in the document]
Comment (tlmquintino) [reply rust/tensogram/src/encode.rs]: Fixed in 69451300. Tightened to `desc.dtype != Dtype::Float64` in `resolve_simple_packing_params` and applied the same tightening to the parallel guard in `build_pipeline_config_with_backend`, so both the auto-compute path and the explicit 4-key path are consistent. The error message now also names the offending dtype. Added BDD test `s8_eight_byte_non_float_dtypes_rejected` exercising `Int64` / `Uint64` / `Complex64` through both the auto-compute path and the explicit 4-key path.
